### DISCUSSION TITLE
protected required properties

### DISCUSSION THREAD

**Discussion by alrz** [dotnet/csharplang#6428] [Language Ideas]
Currently, `SetsRequiredMembers` is required to hint that a member is supposed to set all required members, but in a hierarchy like:
```cs
public abstract class C {
    public required object Prop { get; init; }
}

public sealed class D : C {
    [SetsRequiredMembers]
    public D() { // warns
        // Prop = new();
    }
}
```
There is no way to know that you *should* add it because `D` may be only implicitly created e.g. for a test class.

Marking the setter as `protected` could be a way to enforce the assignment in all derived classes without any additional annotations.
```cs
public abstract class C {
    public required object Prop { get; protected init; }
}

public sealed class D : C {
    public D() { // warns
        // Prop = new();
    }
}
```

**Comment by qrli** [comment]
> There is no way to know that you should add it because D may be only implicitly created e.g. for a test class.

If it is only implicitly created, why would you add `[SetsRequiredMembers]` at all?

**Reply by alrz** [reply]
To get a warning if `Prop = new();` is not there.

**Reply by qrli** [reply]
1st, as I understand, `[SetsRequiredMembers]` is to say all `required` are satisfied for consuming sites, but not to guarantee that the constructor actually does it, except for simple cases. Your purpose is the later, so it is not what it is supposed to do.

2nd, I think, after the ecosystem adopt required properties, implicit creation would also require following initializer (with feature of initializer for factory method), or least would throw if `[SetsRequiredMembers]` is not there.

**Reply by alrz** [reply]
I mentioned SetRequiredMembers as a workaround for now, ~but I think protected members is a viable solution to enforce the assignment in derived classes, because they are not accessible in initializers.~ Scratch that. This isn't the case unless you're doing `new C`.
